src/wamr/filesystem.cpp

```diff
@@ -1,20 +1,432 @@
+    }
+
+    VALIDATE_NATIVE_ADDR(prestatApp, sizeof(wasi_prestat_app_t));
+
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+
+    prestatApp->pr_type = fileDesc.wasiPreopenType;
+    prestatApp->pr_name_len = fileDesc.getPath().size();
+
+    return __WASI_ESUCCESS;
+}
+
+typedef struct iovec_app
+{
+    uint32_t buffOffset;
+    uint32_t buffLen;
+} iovec_app_t;
+
+static int32_t wasi_fd_read(wasm_exec_env_t exec_env,
+                            int32_t fd,
+                            iovec_app_t* ioVecBuffApp,
+                            int32_t ioVecCountApp,
+                            int32_t* bytesRead)
+{
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    storage::FileSystem& fileSystem = module->getFileSystem();
+    std::string path = fileSystem.getPathForFd(fd);
+
+    SPDLOG_DEBUG("S - fd_read {} ({})", fd, path);
+
+    storage::FileDescriptor fileDesc = fileSystem.getFileDescriptor(fd);
+
+    // Translate app iovecs to native ones
+    auto ioVecBuffNative = new ::iovec[ioVecCountApp];
+    for (int i = 0; i < ioVecCountApp; i++) {
+        VALIDATE_APP_ADDR(ioVecBuffApp[i].buffOffset,
+                          sizeof(char) * ioVecBuffApp[i].buffLen);
+
+        ::iovec ioVecNative{
+            .iov_base = ADDR_APP_TO_NATIVE(ioVecBuffApp[i].buffOffset),
+            .iov_len = ioVecBuffApp[i].buffLen,
+        };
+
+        ioVecBuffNative[i] = ioVecNative;
+    }
+
+    // Read from fd
+    VALIDATE_NATIVE_ADDR(bytesRead, sizeof(int32_t))
+    *bytesRead = ::readv(fileDesc.getLinuxFd(), ioVecBuffNative, ioVecCountApp);
+
+    delete[] ioVecBuffNative;
+
+    return __WASI_ESUCCESS;
+}
+
+static int32_t wasi_fd_readdir(wasm_exec_env_t exec_env,
+                               int32_t a,
+                               int32_t* b,
+                               char* c,
+                               int64_t d,
+                               int32_t e)
+{
+    SPDLOG_DEBUG("S - fd_readdir");
+    throw std::runtime_error("fd_readdir not implemented");
+    return 0;
+}
+
+static int32_t wasi_fd_seek(wasm_exec_env_t exec_env,
+                            int32_t fd,
+                            int64_t offset,
+                            int32_t whence,
+                            __wasi_filesize_t* newOffset)
+{
+    SPDLOG_DEBUG("S - fd_seek {} {} {}", fd, offset, whence);
+
+    WAMRWasmModule* module = getExecutingWAMRModule();
+    VALIDATE_NATIVE_ADDR(newOffset, sizeof(__wasi_filesize_t));
+
+    storage::FileDescriptor& fileDesc =
+      module->getFileSystem().getFileDescriptor(fd);
+    auto wasiErrno = fileDesc.seek(offset, whence, newOffset);
+
+    return wasiErrno;
+}
+
+static int32_t wasi_fd_write(wasm_exec_env_t exec_env,
+                             int32_t fd,
+                             iovec_app_t* ioVecBuffApp,
+                             int32_t ioVecCountApp,
+                             int32_t* bytesWritten)
+{
```

Comment:
This method looks incredibly similar to the WAVM version, and it seems like there's not much WAMR/ WAVM specific stuff in it, could we capture some of the logic in the `FileSystem` class instead to avoid duplication?

Reply:
This PR is already very cluttered so I suggest I do this in a separate PR.
Of course, the argument goes both ways, and given how cluttered this PR is I could also do it here.

In favour of the former, there are a couple more `TODO`s in the main PR description that I must do regardless, so it's not likely that I will forget about this at all.

Reply:
Ok cool, separate PR makes sense.